myhpom/models/document.py

```diff
@@ -80,3 +84,86 @@ def remove_documents_on_delete(sender, instance, using, **kwargs):
+    and that are non-guessable. The DocumentUrl model holds "keys" to documents.
+    * advancedirective = foreign key to the AdvanceDirective
+    * key = the non-guessable string that identifies this DocumentUrl
+    * expiration = (optional) timestamp, based on a new setting (now + settings.DOCUMENT_URLS_EXPIRE_IN)
+    * ip = (optional) IP / range to limit IP address to client. Can be:
+        * a single IP address, e.g., "10.16.239.82"
+        * a comma-delimited string with 2 values, e.g., "10.16.239.82, 10.16.239.85"
+        * an IP address with netmask, e.g., "10.16.239.82/24"
+      (uses the iptools.IpRange object)
+    """
+
+    advancedirective = models.ForeignKey(
+        AdvanceDirective,
+        on_delete=models.CASCADE,
+        help_text="The AdvanceDirective to which this URL points.",
+    )
+    key = models.CharField(
+        max_length=48,
+        unique=True,
+        help_text="The non-guessable string that indentifies this DocumentUrl.",
+    )
+    expiration = models.DateTimeField(
+        null=True,  # optional
+        blank=True,
+        help_text="The optional timestamp indicating when this DocumentUrl expires.",
+    )
+    ip = models.CharField(
+        max_length=64,
+        null=True,  # optional
+        blank=True,
+        help_text="The optional IP address or range to which this DocumentUrl is limited",
+        verbose_name='IP',
+    )
+
+    def __str__(self):
+        return self.url
+
+    @property
+    def url(self):
+        return reverse('myhpom:document_url', kwargs={'key': self.key, 'filename': self.filename})
+
+    @property
+    def filename(self):
+        return self.advancedirective.filename
+
+    @property
+    def ip_range(self):
+        """return an IpRange object representing this instance's ip attribute"""
+        if self.ip:
+            return iptools.IpRange(*[ip.strip() for ip in self.ip.split(',')][:2])
+
+    def authorized_client_ip(self, ip_address):
+        """The given ip_address is valid if either:
+        * the instance has no ip_range; or
+        * the ip_address is in this instance's ip_range
+        """
+        if not self.ip_range:
+            return True
+        else:
+            return ip_address in self.ip_range
+
+
+def document_url_before_save(sender, instance, using, **kwargs):
+    """
+    * automatically calculate the key for new instances
+    * automatically set the expiration timestamp for new instances
+        (note: not overloading the expires attribute, since it is expects as a timestamp)
+    """
+    if not instance.pk:
+        # protect against a (very-remotely-possible) key collision
+        while not instance.key or sender.objects.filter(key=instance.key).exists():
+            instance.key = base64.urlsafe_b64encode(str(uuid.uuid4()))
+        if not instance.expiration:
+            instance.expiration = now() + settings.DOCUMENT_URLS_EXPIRE_IN
```

Comment:
I don't think a pre save hook is needed for either of these default field values - use default on the corresponding fields (see comment above). A check against other keys is also not needed - I think you can safely assume uuid is gonna be globally unique!

Reply:
Yeah, you're right.